synthesizer/src/vm/verify.rs

```diff
@@ -449,6 +407,64 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
     }
 }
 
+fn ensure_deployment_valid_for_consensus_version<N: Network>(
+    consensus_version: ConsensusVersion,
+    deployment: &Deployment<N>,
+    id: &N::TransactionID,
+) -> Result<()> {
+    // If the `CONSENSUS_VERSION` is less than `V8`, ensure that
+    //   - the deployment edition is zero.
+    // If the `CONSENSUS_VERSION` is less than `V9` ensure that
+    //   - the deployment edition is zero or one.
+    //   - the program checksum is **not** present in the deployment,
+    //   - the program owner is **not** present in the deployment
+    //   - the program does not use constructors, `Operand::Checksum`, `Operand::Edition`, or `Operand::ProgramOwner`
+    // If the `CONSENSUS_VERSION` is greater than or equal to `V9`, then verify that:
+    //   - the program checksum is present in the deployment
+    //   - the program owner is present in the deployment
+    // If the `CONSENSUS_VERSION` is less than `V10`, then verify that:
+    //   - the program does not use the external struct syntax `some_program.aleo/StructT`
+    if consensus_version < ConsensusVersion::V8 {
+        ensure!(
+            deployment.edition().is_zero(),
+            "Invalid deployment transaction '{id}' - edition should be zero before `ConsensusVersion::V8`",
+        );
+    }
+    if consensus_version < ConsensusVersion::V9 {
+        ensure!(
+            deployment.edition() <= 1,
+            "Invalid deployment transaction '{id}' - edition should be zero or one for before `ConsensusVersion::V9`"
+        );
+        ensure!(
+            deployment.program_checksum().is_none(),
+            "Invalid deployment transaction '{id}' - should not contain program checksum"
+        );
+        ensure!(
+            deployment.program_owner().is_none(),
+            "Invalid deployment transaction '{id}' - should not contain program owner"
+        );
+        ensure!(
+            !deployment.program().contains_v9_syntax(),
+            "Invalid deployment transaction '{id}' - program uses syntax that is not allowed before `ConsensusVersion::V9`"
+        );
+    }
+    if consensus_version >= ConsensusVersion::V9 {
+        ensure!(
+            deployment.program_checksum().is_some(),
+            "Invalid deployment transaction '{id}' - missing program checksum"
+        );
+        ensure!(deployment.program_owner().is_some(), "Invalid deployment transaction '{id}' - missing program owner");
+    }
+    if consensus_version < ConsensusVersion::V10 {
+        ensure!(
```

Comment:
Sanity check. Is there any program that was not deployable in V9 but is deployable in V10 but doesn't explicitly use external structs?

Reply:
External structs is the only addition in terms of program structure.

Additionally, the new fees will be strictly lower so a deployment with too low fee will be accepted from V10, but that's checked elsewhere in the currently open fee PR.

Reply:
@mikebenfield we need to move this check out of `Stack::initialize_and_check` and into `ensure_deployment_valid_for_consensus_version` in order to ensure this is only invalidated from the new consensus version onwards.

Reply:
Right, somehow I misread what you wrote previously. Alright I've moved it now.